### PR TITLE
Improve cgroup device rule validation

### LINKED ISSUE #41920: Improve validation and handling of `--device-cgroup-rule`
As a follow-up to https://github.com/moby/moby/pull/41919

Validation, and handling of `--device-cgroup-rule` could use some improvements.

### 1. Improve handling/validation of "a (all)" value

We currently treat `a` the same as other (`c (char)`, `b (block)`) devices. However, the [kernel documentation](https://github.com/torvalds/linux/blob/v5.10/Documentation/admin-guide/cgroup-v1/devices.rst) describes:

```markdown
... A device cgroup associates a device access whitelist with each cgroup. A whitelist
entry has 4 fields. 'type' is a (all), c (char), or b (block). 'all' means it applies
to all types and all major and minor numbers. Major and minor are either an integer
or * for all. Access is a composition of r (read), w (write), and m (mknod).
```

In addition, that documentation shows an example where _only_ the device-type is passed;

```console
echo a > /sys/fs/cgroup/1/devices.allow
```

Related source code: https://github.com/torvalds/linux/blob/v5.10/security/device_cgroup.c#L614-L642

From the above;

- passing only `a` should is allowed
- passing only `a` is equivalent to "all" device-types, "all" major/minor numbers, and "all" permissions, so equivalent to `a *:* rwm`
- when passing `a` in combination with `major:minor` and `[rwm]` permissions, both major/minor, and permissions are ignored

I think we should:

- consider allowing `a` as valid value
- in that case, either;
    - ignore `major:minor` and `[rwm`, and warn that these fields are ignored (?)
    - produce an error if the fields are not `*:*` and `rwm` respectively (note: permissions can be ordered in any way, so `rwm` and `mwr` are equivalent)

### 2. Improve validation of major/minor

Looks like we currently don't validate against if the major/minor is in the correct range; https://github.com/torvalds/linux/blob/v5.10/include/linux/kdev_t.h
(12 bits are used to encode a major number (0..4095, it seems), 20 bits); https://stackoverflow.com/a/14849363

### 3. Validate on container "create" instead of on "start"

Validation doesn't take place until the container is started; we should validate the value when creating the container, so that we can fail early.

### 4. Update CLI accordingly, or remove client-side validation

We shouldn't validate this value on the client side, and just delegate validation to the daemon/API; if we _do_ want to keep client-side validation, the rules should match what the daemon accepts (if we update).

### PR DESCRIPTION
**Pull request by LeviHarrison**
**- What I did**

Fixes #41920

- Passing only `a` is now allowed
- When one passes only `a`, a cgroup rule allowing all permissions is returned
- If an `a` type rule is passed but is not `a *:* rwm` (`rwm` can be reordered), an error is returned informing the user that the rest of their rule is ineffective ([reference](https://github.com/torvalds/linux/blob/v5.10/Documentation/admin-guide/cgroup-v1/devices.rst))
- Changed the validation of `major` and `minor` numbers so that they are 12 and 20 bits respectively
- Changed the behavior of all (`*`) major or minor numbers to just omit the field in the cgroup permission, as per the [runtime-spec](https://github.com/opencontainers/runtime-spec/blob/v1.0.2/config-linux.md#device-whitelist)
- Updated `oci` tests to reflect these changes
- Moved cgroup validation to container create, instead of container start
- Changed the name of the `AppendDevicePermissionsFromCgroupRules` function to `DevicePermissionsFromCgroupRules` and changed the parameters to reflect that it doesn't need to do any appending anymore
- Added an `errdef` for an error returned by the `DevicePermissionsFromCgroupRules` function and added tests for that `errdef`

One more thing the issue calls for is either updating the CLI or removing the client-side validation. This seems like a decision that has to be made, so I decided to leave it out of this PR for now. I'd be happy to add it once an agreement has been reached.

Also, I just noticed that the regex allows rules such as `c 1:1 rrr` and `b 1:1 wmm`. Should it?

**- How I did it**

- To validate the `a` type rules, I added a function that checks the rule to see if it is either a singular `a` or, for example, a `a 1:5 w`. I then added a second regex that parses the rule to make sure that it's either `a` or `a *:* rwm` (`rwm` can be reordered), and throw an error telling the user that their rule is ineffective if the regex doesn't pass.
- Previously, the cgroup rule validation took place in the `WithDevices` function, which is called through `daemon.createSpec`, which is called on container start. `AppendDevicePermissionsFromCgroupRules` is passed a field from the config (`HostConfig.DeviceCgroupRules`), which is actually set on container create. So, I moved `AppendDevicePermissionsFromCgroupRules` to right after that field is set in container create, and set the result of it to a new field (`HostConfig.DeviceCgroupPermissions`). Because that `WithDevices` function has access to the config, I replaced the block of code that ran the function with one that just appends `HostConfig.DeviceCgroupPermissions` to the accumulated `devPermissions`. A similar pattern exists in `oci_windows.go`, and I did the same there. Because this function is now not appending to any existing value, just being used to set a field, I decided to change the name to `DevicePermissionsFromCgroupRules` and remove the parameter that provided the existing value.

**- How to verify it**

- Run `oci` tests

```shell
curl -v --unix-socket /var/run/docker.sock \
"http://localhost/v1.41/containers/create?name=foobar" \
-H "Content-Type: application/json" \
-d '{"Image":"hello-world", "HostConfig":{"DeviceCgroupRules": ["invalid rule"]}}'
```

`{"message":"invalid device cgroup rule format: 'invalid rule'"}`

**- Description for the changelog**

`Improved device cgroup rule parsing and moved it to container create`

**- A picture of a cute animal (not mandatory but encouraged)**

image

### CODE REVIEW

**Comment by LeviHarrison** [comment]
Not sure what's going on with the `amd64` tests here. All the failure points work fine on my machine, and it fails at different places on different runs for seemingly not reason.

**Comment by thaJeztah** [comment]
CI failures look unrelated and are known to be flaky (https://github.com/moby/moby/issues/36907 and https://github.com/moby/moby/issues/41561)

File: oci/oci.go
Comment (thaJeztah) [inline oci/oci.go:50]: w.r.t. Major/Minor `nil` being equivalent to `-1` (runtime-spec); Looks indeed that runc uses these as a default, and only override in the "not nil" situation; https://github.com/opencontainers/runc/blob/ee06c700dcb34504204cb171bb35e0e01477bf58/libcontainer/specconv/spec_linux.go#L475-L487

Should we check if other runtimes (e.g. `crun` handle it the same?)

File: oci/oci.go
Comment (thaJeztah) [inline oci/oci.go:68]: I'm wondering if `12 bit` is correct here. I recall the kernel has some "weird" things; e.g., using an "unsigned" integer, but allowing `-1` as "special value". In this case (just from a quick look); it looks like it's indeed using `12 bit unsigned` and `20 bit unsigned` for these; https://github.com/torvalds/linux/blob/139711f033f636cc78b6aaf7363252241b9698ef/include/linux/kdev_t.h#L7-L11

Which would mean (`-1`?) _or_ `0-4096` / `0-1048576` ?

While definitely not the "canonical source of truth", this StackOverflow question has an interesting quote as well; https://stackoverflow.com/a/14849363/1811501 (emphasis mine)

> As of Version 2.6.0 of the kernel, dev_t is a 32-bit quantity with 12 bits set aside for the major number and 20 for the minor number. ***Your code should, of course, never make any assumptions about the internal organization of device numbers; it should, instead, make use of a set of macros found in linux/kdev_t.h***

File: oci/oci_test.go
Comment (thaJeztah) [inline oci/oci_test.go:91]: See my other comment; we'd have to check if 2048 is indeed out of range (it may not be).

Possibly we may have to skip range validation (other than conversion to the `LinuxDeviceCgroup.Major` / `LinuxDeviceCgroup.Minor`), and leave it to the OCI runtime and the Kernel itself to decide wether or not the value is "ok"

**Review by thaJeztah — CHANGES_REQUESTED** [review]
Thanks! Had a first pass at your changes, and left comments/thoughts/suggestions inline while looking over them

File: oci/oci.go
Comment (LeviHarrison) [reply oci/oci.go]: `crun` does seem to [use `-1` as their default/everything value](https://github.com/containers/crun/blob/ebba51f329cd7e3b508febb1e9752b573f19f6a4/src/libcrun/cgroup.c#L2435). 

When I wrote this, I was referencing the [runtime-spec](https://github.com/opencontainers/runtime-spec/blob/v1.0.2/config-linux.md#device-whitelist), which describes that unset major/minor numbers map to the default `*`. The same goes for the Type and Access too, but that's irrelevant. For the sake of conformity and explicitness though, I'll add those defaults back.

File: daemon/oci_linux.go
Comment (LeviHarrison) [reply daemon/oci_linux.go:915]: I never thought about that aspect, thanks! I would put forth adding a migration step to avoid either adding more logic to the validation function or adding a second (less strict) one, although I can see why having those kinds of things around could be undesirable. Up to you.

File: oci/oci.go
Comment (LeviHarrison) [reply oci/oci.go:68]: Yeah, that does seem to be the case.

File: oci/oci_test.go
Comment (LeviHarrison) [reply oci/oci_test.go:91]: Although I think range validation is important (especially because it takes place on `create` now and the Kernel will only check on `start`, which could give a user false confidence in their rules), the general uncertainty surrounding it, with both the `runtime-spec` and `runc` ignoring the limit, gives me pause. Maybe we just return a warning, something along the lines of  `this value probably will be out of range`.

**Comment by LeviHarrison** [comment]
I fixed the major and minor range, errdefs, and added warnings. I'm still waiting on further feedback before I resolve the issue with setting permissions in restarted containers.

**Pull request closed without merging**
